README.md
Pesterchum Online
================
This is the repository for Pesterchum Online, a browser-based client for the Pesterchum IRC network. It runs off NodeJS, using the NodeIRC library. Express and EJS are used for serving pages. All code in this repository is licensed under the GNU General Public License v3 - read more about it [here](https://www.gnu.org/copyleft/gpl.html) or in the [License section](#license) of this readme.

##Getting started  
PCO uses NodeJS - if you haven't installed that yet, [do so](http://nodejs.org/download/).  
Run `npm install` in the root directory to get all the required modules down, then run `node PCO.js` to start the app. If everything went in right, you should see something like this:

```
PCO v0.1.0 started.
  Configured HTML engine using EJS.
  Using HTML as view engine.
  Loaded bodyParser.
  Serving views on /views.
  Serving resources on /public.
  Listening on port 612.
```

The app uses `process.env.PORT` if it exists (if you're using Heroku or Cloud9, for example) - if not, it defaults to `:612`. Open a browser to the logged port to see where everything starts.

##Technical overview  
###Basic functionality:  
The user starts on the index page and inputs a handle. This is checked against Pesterchum specifications using the RegEx `/[a-z0-9]*[A-Z][a-z0-9]*$/` (the handle must not start with a capital letter, have a single capital letter, and contain only alphanumeric characters). Prefixing the handle with override_ will allow overriding the handle validation. The handle is passed to the main page using a URL parameter (?nick=[handle]).

The client goes to the main page and gets the nick. The server double-checks the nick with the same RegEx and override validation and HTTP 502 redirects the client back to the index if the nick doesn't validate. If it does, the client removed the override_ prefix if it's there, then transmits the nick to the server in an HTTP `POST` to /znewclient. The server receives the request and does the following:

* Increments the `clientstotal` counter
* Pushes an object to the `clients` array containing the client's connection details
* Sets an interval corresponding to the new client's ID that increments the client's `missedpings` every 0.75 seconds
    * If five pings have been missed, the client is killed and the interval is cleared
* Creates an array for the client in the `clientlogs` array
* Initiates an IRC connection to `irc.mindfang.org` using the client's object in the `clients` array
* Adds a `message` listener to the connection
    * Adds a space at the end of the message if one doesn't exist already (this solves issues with IRC and Chumdroid)
    * Replaces color tags and special characters using `htmlFormatFct()`
    * Checks if the channel the message was sent from is `#pesterchum` or if the message is a `PESTERCHUM:TIME` message
        * If not, formatted message is pushed to the `clientlogs` array corresponding to the client
        * If so, **[WIP]**
* Sends the client its object from the `clients` array

The client sets its global `client` variable to the received object and moves on. It then sets an interval to run `ircUpdateFct()` every 0.75 seconds to match the server's ping frequency. That function does the following:

* Transmits the client's ID to the server in an HTTP `POST` to /zupdate
    * The server resets the client's `missedpings` to 0 and sends back its `clientlogs` array
* If the `POST` fails (i.e. the server is offline/not responding), the client kills itself and reloads the page
* If the `POST` is successful, it receives its message log and displays it by joining the array items with a `<br>`
* The message display div is scrolled to the bottom using `scrollTop` and `scrollHeight`

The client then prepares all UI content and event handlers.

###Active functionality:  
**Joining a memo:**  
The client checks to see if the memo already exists in its `channels` array (i.e. if it is already in the memo). If not, it transmits its ID and the requested memo to the server in an HTTP `POST` to /zjoinememo. The server uses the client's ID to have the corresponding IRC connection send a channel join request.

**Sending a message:**  
The client transmits its ID, the message, and the memo to which to send the message to the server in an HTTP `POST` to /zsendmessage. The server receives this and does the following:

* Uses `handle[0].toUpperCase()+/[A-Z]/.exec(handle)[0]` to get the client's two-letter prefix
* Adds the prefix to the message along with its color
* Uses the client's ID to have the corresponding IRC connection send the message to the memo
* Replaces color tags and special characters using `htmlFormatFct()`
* Pushes the formatted message to the `clientlogs` array corresponding to the client

##Development guidelines  
###Writing code  
Do your best to maintain consistency with the rest of the code - indents are done with four spaces, all same-line comments are indented to column 81, and `"use strict";` should be used everywhere. If you're not sure about how to do something from a stylistic standpoint, [submit an issue](https://github.com/Hydrothermal/PesterchumOnline/issues/new) with the question tag or just go what you feel is best and it'll be fixed during the next code cleanup. Use the `debug.airplane` switch to work on non-connection things without actually connecting to the server.

###Committing and pushing 
Push all work to the `dev` branch - `master` is for releases only. Release version numbering follows a MAJOR.MINOR.PATCH model, roughly defined below:

* **Major version**: A release involving multiple features and significant changes - this is likely to break any third-party programs that relied on the old structure. Heavy testing should go into the entire application before releasing a major version.
* **Minor version**: A release involving one or few features/changes, and not likely to break third-party programs. At least heavy unit testing should go into the areas changed in a minor version release.
* **Patch**: A bugfix or extremely minor feature unlikely to break anything or change any sections of the application outside of the patched areas. Heavy testing is probably not necessary, especially because most patches should be rolled out as quickly as possible.

When pushing work, make sure to include a concise description of your changes in your commit message. If it seems like a good idea to expound on your changes or explain why they were made, add a comment to the commit. If your changes are significant enough to be important to other developers or tucked-away enough that it's not immediately obvious, updated the technical overview in `README.md` with the appropriate information. Include your changes in the changelog section for the current version when pushing.

**Do not ever use `--force`/`-f` when pushing a commit unless you are forcing your own commit AND are doing it within under a minute of the original push. It's understandable that forcing a push sometimes comes in handy if you're fixing a very minor mistake that you already pushed, but doing so when your commit has already been up for a while gets confusing, and overwriting somebody else's commit is right out. Review your code thoroughly before pushing.**

###License  
Pesterchum Online uses the GNU GPL-3 license - basically, you can copy, distribute, and modify any code as long as you track your changes, license all redistributions under GPL, and link back to the original ([this repo](https://github.com/Hydrothermal/PesterchumOnline)). You can distribute commercially, but you have to keep the source code open. You can't hold me or any other developers liable for any damages you incur related to the code or this project.

####This readme is a WIP and will be updated as progress is made.

CHANGELOG.md
1.2.2 (07/06/2014)
==================
Features:

* Changed client IDs from incremental to random

1.2.1 (05/06/2014)
==================
Features:

* Added messages for users joining and parting from memos

1.2.0 (05/06/2014)
==================
Features:

* Changing color and nick are now possible

Bugfixes:

* Fixed the client rejoining memos after being kicked

1.1.1 (21/05/2014)
==================
Features:

* Color square for indicating text color

Bugfixes:

* Fixed the server crashing when the client made calls too quickly

1.1.0 (05/05/2014)
==================
Features:

* Internal changes (cleaner CSS; general cleanup)
* Style overhaul of chat page
* Added license and changelog

1.0.0 (03/05/2014)
==================
Features:

* Client can enter a handle, join memos, and send messages to memos currently present in
* Client receives all messages sent to memos currently present in (PESTERCHUM:TIME and #pesterchum messages are squelched)
* Server creates new client connections on request and handles message logs for each one
* Client connections are terminated after 3.75 seconds of not requesting a ping update
* HTML formatting converts color code tags and text URLs into colored spans and clickable hyperlinks, respectively

public/client_core.js
var client = {},                                                                //Client object
    forcequit = false,                                                          //Forcequit variable
    curchan = "#PesterchumOnline";                                              //Selected channel

function ircUpdateFct() {
    "use strict";
    var histel, memolist;
    $.post('./zupdate', {id:client.id}, function(data){                         //Request an update from the server
        $('#content').html(data[0].join("<br>"));                               //Load in this client's log
        client = data[1];                                                       //Load in this client's object
        memolist = "";                                                          //String to hold dropdown HTML
        for(var i = 0; i < client.channels.length; i++) {                       //Loop through all channels
            if(client.channels[i] === "#pesterchum") { continue; }              //Skip #pesterchum
            if(client.channels[i] === curchan) {
                memolist += "<option selected>" + client.channels[i] + "</option>"; //Option tag for the current channel (selected)
            } else {
                memolist += "<option>" + client.channels[i] + "</option>";      //Option tag for the other channels
            }
        }
        $("#memoselect").html(memolist);                                        //Put option tag HTML into the dropdown
        $("#colorsquare").css("background-color", "rgb(" + client.color + ")"); //Text color
        $("#mynick").html(client.nick);                                         //Nick
    }).error(function() {                                                       //If the server isn't responding
        forcequit = true;                                                       //Force a quit
        location.reload();                                                      //Reload to try and request the page again
    });
    histel = document.getElementById('content');                                //Chat log
    histel.scrollTop = histel.scrollHeight;                                     //Scroll to the bottom
}

function joinMemoFct(memo) {
    "use strict";
    if(client.channels.indexOf(memo)===-1) {                                    //Make sure you're not already in the requested memo
        $.post('./zjoinmemo', {id:client.id, memo:memo}, function(data){        //Request a memo join from the server
            console.log("Joined "+memo);                                        //Success
        });
    }
}

function sendMessageFct(memo,message) {
    "use strict";
    $.post('./zsendmessage', {id:client.id, memo:memo, message:message}, function(data){ //Request a message send from the server
        console.log("Sent message "+message+" to "+memo);                       //Success
    });
}

window.onload = function() {
    "use strict";
    var nick = document.URL.split("?nick=")[1],
        updateInterval;
    if(nick.substr(0,9) === "override_") {
        nick = nick.substr(9);
    }

    $("#mynick").html(nick);                                                    //Your current nick
    $.post('./znewclient', {                                                    //Request a new client from the server
        nick: nick,
        color: "255,0,0"                                                        //Hardcoded color for now
    }, function(data){
        var rgb, r, g, b, hex;
        client = data;                                                          //Fill the returned client details into the client object
        ircUpdateFct();                                                         //Update

        //Convert RGB text color to hex
        rgb = client.color.split(",");                                          //Split commas
        r = parseInt(rgb[0], 10);                                               //Convert to integer
        g = parseInt(rgb[1], 10);                                               //Convert to integer
        b = parseInt(rgb[2], 10);                                               //Convert to integer
        hex = "#" + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1); //Bitwise ops and hash prefix
        $("#colorsquare").css("background-color", hex);                         //Set colorsquare
        
        $("#colorsquare").ColorPicker({
            color: hex,
            onShow: function (colpkr) {
                $(colpkr).fadeIn(150);
                return false;
            },
            onHide: function(colpkr) {
                $(colpkr).fadeOut(150);
                return false;
            },
            onSubmit: function(hsb, hex, rgb, obj, colpkr) {
                var rgb = rgb.r + "," + rgb.g + "," + rgb.b;
                $("#colorsquare").css("background-color", "rgb(" + rgb + ")");
                $(colpkr).fadeOut(150);
                $.post('./zchangecolor', {
                    id: client.id,
                    color: rgb
                });
                ircUpdateFct();                                                 //Force update
                return false;
            }
        });

        $("#mynick").click(function() {
            var newnick = prompt("New nick:");                                  //Change this to a floaty div thing at some point
            if(!newnick) { return false; }                                      //Stop if you clicked cancel
            var valhandle = /^[a-z0-9]*[A-Z][a-z0-9]*$/.test(newnick),          //Validate the new nick
                override = newnick.substr(0,9) === "override_";                 //Override prefix
            
            if(valhandle || override) {
                $.post('./zchangenick', {
                    id: client.id,
                    nick: newnick
                }, function(success) {
                    if(success) {
                        ircUpdateFct();
                    } else {
                        //Rejection alert
                        alert("That is not a valid chumhandle.\nA chumhandle must not start with a capital letter, have a single capital letter, and contain only alphanumeric characters.");
                    }
                });
            } else {
                //Rejection alert
                alert("That is not a valid chumhandle.\nA chumhandle must not start with a capital letter, have a single capital letter, and contain only alphanumeric characters.");
            }
        });
    });
    
    updateInterval = setInterval(function(){                                    //Automatic update
        ircUpdateFct();                                                         //Update
    },750);                                                                     //.75 seconds
    
    $("#joinsend").click(function() {                                           //Join button
        joinMemoFct($("#joinfield").val());                                     //Join the memo
        $("#joinfield").val("");                                                //Empty the join input
    });
    
    $("#send").click(function() {                                               //Send button
        sendMessageFct(curchan,$("#field").val());                              //Send the message
        $("#field").val("");                                                    //Empty the message input
    });

    $("#memoselect").change(function() {
        curchan = $("#memoselect").val();                                       //Get the selected memo from the dropdown
    });
};

window.onbeforeunload = function() {                                            //Catch closing the page
    "use strict";
    if(forcequit === false) {                                                   //If we're not forcing a quit
        return "If you leave this page, you will disconnect from the server and close all memos/pesters."; //Request confirmation
    }
};

PCO.js
//Environment setup
var port = +process.env.PORT || 612,                                           //Use :612 if we're testing locally; environment port if we're live
    express = require("express"),
    irc = require("irc"),
    bodyParser = require("body-parser"),
    ejs = require('ejs'),
    pjson = require("./package.json"),
    app = express(),
    crypto = require('crypto'),

//Declare some global variables
    clientstotal = 0,                                                          //Client ID counter
    clients = {},                                                              //Client data objects
    connections = [],                                                          //IRC client objects
    clientlogs = [],                                                           //Client message logs
    pingchecks = [],                                                           //Ping update intervals
    Pesterchum = {                                                             //Pesterchum helper object
        Messages: {}                                                           //Message emitters
    },
    debug = {
        airplane: false,
        suppressed: false
    },
    ip;

function applog(text) {
    "use strict";
    if(!debug.suppressed) { console.log("  " + text) };
}

function killClientFct(id, reason) {
    "use strict";
    if(reason===undefined) { reason = "Quit"; }                                //If you didn't specify a reason, assume it was a normal quit
    if(!debug.airplane) { connections[id].disconnect(reason); }                //Disconnect the client with the specified reason
    applog("Killed client " + CIDFct(id) + " for reason " + reason + "."); //Log
}

function htmlFormatFct(message) {
    "use strict";
    var urlregex = /((?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:\/[^\s<]*)?)/i;
    return message.replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/>/g,"&gt;")  //Protect from XSS
                  .replace(/&lt;[cC]=(\d{1,3},\d{1,3},\d{1,3})&gt;/g,"<c=rgb($1)>") //Convert RGB color codes to CSS RGB syntax
                  .replace(/&lt;[cC]=([^&]*)&gt;/g,"<span style='color: $1'>")      //Convert all color tags to spans
                  .replace(/&lt;\/c&gt;/g,"</span>")                                //Convert closing color tags to closing spans
                  .replace(/\s{2,}/g," ")                                           //Remove extra spaces
                  .replace(urlregex, "<a href='$1'>$1</a>");                        //Replace URls with links
}

function getIPFct(req) {
    "use strict";
    return req.headers['x-forwarded-for'] || req.connection.remoteAddress;
}

function denyRequestFct(reason) {
    "use strict";
    applog("Request denied (" + reason + ")");
}

function pad(num, len) {
    "use strict";
    return ("000000000000000" + num).slice(-len);                              //Add a bunch of zeroes to the front and slice from the end
}

function CIDFct(id) {
    return "[" + id.substr(0, 8) + "] (" + clients[id].nick + ")";
}

Pesterchum.getPrefixFct = function(handle) {
    "use strict";
    return handle[0].toUpperCase() + /[A-Z]/.exec(handle)[0];                  //Return the handle's prefix
};

Pesterchum.validateHandleFct = function(handle) {
    "use strict";
    return (/^[a-z0-9]*[A-Z][a-z0-9]*$/).test(handle);
};

Pesterchum.Messages.time = function(time) {
    "use strict";
    time = time || "?";                                                        //Default to ??:??

    if(time === 0) {
        return "PESTERCHUM:TIME>i";                                            //Current time
    }

    if(time > 0) {
        time = pad(time, 4);                                                   //Pad to four digits
        time = time.substr(0, 2) + ":" + time.substr(2);                       //Insert a colon into the middle
        return "PESTERCHUM:TIME>F" + time;                                     //Future time
    }

    if(time < 0) {
        time = pad(-time, 4);                                                  //Flip from negative to positive and pad to four digits
        time = time.substr(0, 2) + ":" + time.substr(2);                       //Insert a colon into the middle
        return "PESTERCHUM:TIME>P" + time;                                     //Past time
    }

    return "PESTERCHUM:TIME>?";                                                //Unknown time
};

Pesterchum.Messages.mood = function(mood) {
    "use strict";
    mood = mood || "chummy";                                                   //Default to chummy
    //Taken directly from Pesterchum's mood.py file
    var moods = ["chummy", "rancorous", "offline", "pleasant", "distraught",
                 "pranky", "smooth", "ecstatic", "relaxed", "discontent",
                 "devious", "sleek", "detestful", "mirthful", "manipulative",
                 "vigorous", "perky", "acceptant", "protective", "mystified",
                 "amazed", "insolent", "bemused" ],
                 moodnum = moods.indexOf(mood.toLowerCase());                  //Get the index of the requested mood

    if(moodnum !== -1) {                                                       //If the requested mood matched
        return "MOOD >" + moodnum;                                             //Return the mood message with the requested mood's index
    }

    //If the requested mood did not match
    applog("Requested mood " + mood + " not found; defaulting to chummy.");    //Log the mistake
    return "MOOD >0";                                                          //Default to chummy
};

if(!debug.suppressed) {
    console.log("PCO v" + pjson.version + " started.");                            //Log startup
}

//Setting stuff up
app.engine('htm', ejs.renderFile);
    applog("Configured HTML engine using EJS.");
app.set('view engine', "html");
    applog("Using HTML as view engine.");
app.use(bodyParser());
    applog("Loaded bodyParser.");
app.set('views', __dirname + '/views');
    applog("Serving views on /views.");
app.use(express.static(__dirname + '/public'));
    applog("Serving resources on /public.");
app.listen(port);
    applog("Listening on port " + port + ".");

//Listeners
app.get("/", function(req, res){                                               //Index
    "use strict";
    res.render("index.htm");                                                   //Render index.htm
    ip = getIPFct(req);                                                        //Client IP address
    applog("Rendered index.htm for " + ip + ".");                              //Log
});
app.get("/chat", function(req, res){                                           //Chat page
    "use strict";
    var nick = req.param("nick"),                                              //Get the requested nick
        override = nick.substr(0,9) === "override_";                           //Override prefix

    if(!Pesterchum.validateHandleFct(nick) && !override) {                     //Check to see if the handle fails validation
        ip = getIPFct(req);                                                    //Client IP address
        applog("Rejected invalid handle request from " + ip + ".");            //Log rejection
        res.redirect("/");                                                     //HTTP 502 back to the index page
        return false;                                                          //Stop processing
    }

    res.render("chat.htm");                                                    //Render chat.htm
    ip = getIPFct(req);                                                        //Client IP address
    applog("Rendered chat.htm for " + ip + ".");                               //Log
});

app.post("/zupdate", function(req, res){                                       //Update interval
    "use strict";
    var clientid = req.body.id;                                                //Get the client's ID
    clients[clientid].missedpings = 0;                                         //Reset the client's missed pings to 0
    res.send([clientlogs[clientid], clients[clientid]]);                       //Give the client its chatlog and its object
});

app.post("/zjoinmemo", function(req, res){                                     //Joining a memo
    "use strict";
    var clientid = req.body.id,                                                //Get the client's ID
        memo;

    if(clients[clientid].ready) {
        if(req.body.memo[0] === "#") {                                         //If you prefixed the memo with a #
            memo = req.body.memo;                                              //Get the requested memo
        } else {                                                               //If you didn't prefix the memo with a #
            memo = "#" + req.body.memo;                                        //Get the requested memo and add the #
        }
        if(!debug.airplane) { connections[clientid].join(memo); }              //Join the requested memo
        clients[clientid].channels.push(memo);                                 //Add the requested memo the client object's channel list
        applog("Client " + CIDFct(clientid) + " joined memo " + memo + ".");   //Log
    } else {
        applog("Client " + CIDFct(clientid) + " attempted to join memo " + memo + "."); //Log
        denyRequestFct("client was not ready");
    }
});

app.post("/zsendmessage", function(req, res){                                  //Sending a message
    "use strict";
    var clientid = req.body.id,                                                //Get the client's ID
        targ = req.body.memo,                                                  //Get the requested target
        message = req.body.message,                                            //Get the message
        handle = clients[clientid].nick,                                       //Get the client's handle
        color = clients[clientid].color,                                       //Get the client's text color
        prefix;

    if(clients[clientid].ready) {
        prefix = Pesterchum.getPrefixFct(handle);                              //Get the client's prefix
        message = "<c=" + color + ">" + prefix + ": " + message + "</c>";      //Compile actual message
        if(!debug.airplane) { connections[clientid].say(targ, message); }      //Send the message
        var htmlmsg = "<span style='font-weight:bold'>" + targ + ": </span>" + message; //HTML Channel prefix - to be removed in favor of tabs
        clientlogs[clientid].push(htmlFormatFct(htmlmsg));                     //Add the message to the client log

        applog("Client " + CIDFct(clientid) + " sent message \"" + message + "\" to memo " + targ + "."); //Log
    } else {
        applog("Client " + CIDFct(clientid) + " attempted to send message \"" + message + "\" to memo " + targ + "."); //Log
        denyRequestFct("client was not ready");
    }
});

app.post("/zchangecolor", function(req, res){                                  //Color change
    "use strict";
    var clientid = req.body.id;                                                //Get the client's ID
    var color = req.body.color;                                                //Get the requested color
    clients[clientid].color = color;                                           //Change the client's color
});

app.post("/zchangenick", function(req, res){                                   //Color change
    "use strict";
    var clientid = req.body.id;                                                //Get the client's ID
    var nick = req.body.nick,                                                  //Get the requested nick
        override = nick.substr(0,9) === "override_";                           //Override prefix

    if(!Pesterchum.validateHandleFct(nick) && !override) {                     //Check to see if the handle fails validation
        ip = getIPFct(req);                                                    //Client IP address
        applog("Rejected invalid handle request from " + ip + ".");            //Log rejection
        res.send(false);                                                       //Tell the client
        return false;                                                          //Stop processing
    }

    if(nick.substr(0,9) === "override_") {
        nick = nick.substr(9);
    }
    
    clients[clientid].nick = nick;                                             //Change the client's nick
    connections[clientid].send("NICK", nick);                                  //Send the nick change to the server
    res.send(true);                                                            //Tell the client
});

app.post("/znewclient", function(req, res){                                    //Initial new client request
    "use strict";
    var nick = req.body.nick,                                                  //Get the requested nick
        id, config, rand;
    ip = getIPFct(req);                                                        //Client IP address
    applog("Responded to /znewclient request from " + ip + ".");               //Log response

    clientstotal += 1;                                                         //Increment the client counter

    id = crypto.randomBytes(32).toString("base64");                            //Cryptographically strong random id

    //If the ID already exists, loop until unique
    for(var i = 0; !!clients[id]; i++) {
        id = crypto.randomBytes(32).toString("base64");
    }

    config = {                                                                 //Create the new client
      "id": id,                                                                //Unique ID
      "nick": nick,                                                            //Handle
      "color": req.body.color,                                                 //Text color
//    "userName": "pco" + id,                                                  //Username using ID
      "userName": "pcc31",                                                     //Spoof Pesterchum client
      "realName": "pco" + id,                                                  //Realname using ID - to be removed in favor of IP address or hostmask
      "missedpings": 0,                                                        //Number of updates missed
      "channels": ["#pesterchum","#PesterchumOnline"],                         //Initial channels
      "ready": false                                                           //Connection ready boolean
    };
    clients[id] = config;
    
    pingchecks[id] = setInterval(function(){                                   //Create a pingcheck interval for the new client
        clients[id].missedpings += 1;                                          //Increment the missed pings by one
        if(clients[id].missedpings >= 5) {                                     //If the client has missed five pings
            killClientFct(id,"Ping timeout");                                  //Kill the client
            clearInterval(pingchecks[id]);                                     //Clear the interval
        }
    }, 750);                                                                   //Checks for a ping every .75 seconds
    
    clientlogs[id] = [];                                                       //Create a log for the client
    
    if(!debug.airplane) {
        connections[id] = new irc.Client("irc.mindfang.org", nick, {           //Connect to the server
            channels: config.channels,
            userName: config.userName,
            realName: config.realName,
            autoRejoin: false
        });
        connections[id].addListener("message", function(from, to, text, message) {
            var channel = message.args[0],                                     //Set the channel
                msgtext = message.args[1],                                     //Set the message text
                lastchar;

            lastchar = msgtext.charAt(msgtext.length - 1);                     //Get the last character of the message
            if(lastchar !== " ") { msgtext += " "; }                           //If it's not a space, add a space. This solves IRC + chumdroid issues.
            
            msgtext = htmlFormatFct(msgtext);
            
            if(channel !== "#pesterchum" && msgtext.search("PESTERCHUM:TIME>")===-1) { //Ignore #pesterchum and PESTERCHUM:TIME messages
                clientlogs[id].push("<span style='font-weight:bold'>" + channel + ": </span>" + msgtext); //HTML Channel prefix - to be removed in favor of tabs
            }
        });
        connections[id].addListener("registered", function() {
            clients[id].ready = true;
        });
        connections[id].addListener("join", function(channel, who) {
            if(channel !== "#pesterchum" && who !== clients[id].nick) {
                clientlogs[id].push(who + " joined " + channel + ".");
            }
        });
        connections[id].addListener("part", function(channel, who) {
            if(channel !== "#pesterchum" && who !== clients[id].nick) {
                clientlogs[id].push(who + " left " + channel + ".");
            }
        });
    }
    
    res.send(config);                                                          //Give the client its object
    applog("Created new client with ID of " + id + " and handle of " + nick + ".");
});
